### PR TITLE
Move to AES-256-GCM for openssl_seal/open

### PR DESCRIPTION
**Pull request by rullzer**
Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

### CODE REVIEW

**Review by nickvergessen — APPROVED** [review]
🐘

File: apps/encryption/lib/Crypto/Crypt.php [hunk 1]
```diff
@@ -677,11 +677,25 @@ public function multiKeyDecrypt($encKeyFile, $shareKey, $privateKey) {
 			throw new MultiKeyDecryptException('Cannot multikey decrypt empty plain content');
 		}
 
+		$prev = null;
+
+		// We need to be able to extract the IV
+		if (strlen($encKeyFile) > 12) {
```
Comment (LukasReschke) [inline apps/encryption/lib/Crypto/Crypt.php:683]: Doesn't this mean we'd always go into this block or do I miss something?  Is there a case in which this is `=< 12` for older files?

Comment (rullzer) [reply apps/encryption/lib/Crypto/Crypt.php:683]: I'm not sure might be for older files yes...

**Comment by J0WI** [comment]
#2182 suggests to move to `openssl_public_encrypt`

**Comment by rotdrop** [comment]
Hi, of course, this is not a PHP lesson, so my apologies if I am doing something obviously wrong. Out of curiosity:

I am not able to use openssl_seal()/openssl_open() with the aes-256-gcm cipher. I suspect that PHP does not take the authentication tag into account which is needed to validate the encrypted data. Specifically, the example code works for aes-256-ctr but not for the gcm cipher. I hope the example is fairly minimal.

For demonstration purposes I have also added a direct call to openssl_decrypt(), which of course cannot work for the gcm-cipher. However, neither does openssl_open(). 

So I wonder why your code is able to decrypt data sealed the gcm cipher?

```
<?php

const NUM_KEYS = 4;

for ($i = 0; $i < NUM_KEYS; $i++) {
  $keys[] = $key = openssl_pkey_new();
  $details = openssl_pkey_get_details($key);
  $pubKeys[] = $details['key'];
  $privKeys[] = openssl_pkey_get_private($key);
}

$data = 'Hello World! Blub';

$ciphers = [
  'aes-256-ctr',
  'aes-256-gcm',
];

foreach ($ciphers as $cipherAlgo) {

  echo "*** TESTING $cipherAlgo ***" . PHP_EOL . PHP_EOL;
  $iv = \random_bytes(openssl_cipher_iv_length($cipherAlgo));
  $result = openssl_seal($data, $sealedData, $sealedKeys, $pubKeys, $cipherAlgo, $iv);

  echo 'DATA  : ' . strlen($data) . ' ' . $data . PHP_EOL;
  echo "IV-LEN: " . openssl_cipher_iv_length($cipherAlgo) . PHP_EOL;
  echo "IV    : " . bin2hex($iv) . PHP_EOL;
  echo 'RESULT: ' . strlen($sealedData) . ' ' . bin2hex($sealedData) . PHP_EOL;
  echo PHP_EOL;

  // Try decrypt
  foreach ($keys as $i => $key) {
    $decrypted = null; // ;)
    openssl_open($sealedData, $decrypted, $sealedKeys[$i], $key, $cipherAlgo, $iv);
    echo "OPEN:    " . $decrypted . PHP_EOL;

    openssl_private_decrypt($sealedKeys[$i], $unsealedKey, $key);
    echo "UNSEAL:  " . bin2hex($unsealedKey) . PHP_EOL;
    echo "DECRYPT: " . openssl_decrypt($sealedData, $cipherAlgo, $unsealedKey, OPENSSL_RAW_DATA, $iv) . PHP_EOL;
  }

  echo PHP_EOL;
}
```

**Comment by solracsf** [comment]
Any move here? This is becoming a problem with recent OS (Ubuntu 22.04, Fedora 36...) shipping OpenSSL v3 which removes RC4 cipher by default. See https://github.com/nextcloud/server/issues/32003

**Comment by come-nc** [comment]
@rullzer @rotdrop @yahesh 
Is there any particular reason to use AES-256-GCM? We could switch to any cipher supported by openssl_seal but not obsolete yet, no? What is the added value of ciphers which needs an auth tag?

**Comment by yahesh** [comment]
@come-nc IMHO cryptographic processing should fail early. Currently, the encryption of the filekeys is not integrity-protected, meaning that an attacker can replace the filekey by knowing the recipients' public keys and Nextcloud will use it until finally the HMAC check of the individually encrypted file blocks fails. Using GCM mode would already fail upfront.

I'd be even more thrilled to get rid of `openssl_seal` and `openssl_open` entirely as it introduces additional intermediate keys don't help to gain any security benefits. But that's a totally different story.

**Comment by rotdrop** [comment]
> A potential solution could be to reimplement the `openssl_seal()` functionality which is nothing more than encrypting `$data` with a random `$key` and then encrypting `$key` with the provided public keys. This would open up a lot of possibilities:
> 
> 1.) A new scheme that uses AES-256-GCM could be introduced. 2.) One could finally switch from PKCS 1.5 padding to OAEP padding. 3.) Updating the share list could be sped up as newly shared files wouldn't have to be re-encrypted anymore (but the intermediate would just have to be encrypted for the added users). 4.) A future path would be opened up to get rid of the intermediate key altogether, which IMHO doesn't serve any real purpose. 5.) A fallback-mechanism could be introduced that uses a PHP implementation of RC4 in case OpenSSL v3 is installed.

... it would even be possible to not use openssl at all and choose something else like Halite. But yes: in order to use autenticated cipher text one has to replace open_ssl() by something similar.  If done properly one could then make the use of openssl an option.

**Comment by yahesh** [comment]
> ... it would even be possible to not use openssl at all and choose something else like Halite. [...]

Which would make it more difficult to install Nextcloud as the `php-sodium` package would be required. OpenSSL support is typically compiled into PHP and is therefore widely available.

**Comment by yahesh** [comment]
With https://github.com/nextcloud/server/pull/35916 I introduced `wrapped_openssl_seal()` and `wrapped_openssl_open()` which contains a reimplementation of `openssl_seal()` and `openssl_open()`. These functions therefore contain everything that's needed to replace the standard functions with something individual which could also properly handle AEAD tags.

**Comment by blizzz** [comment]
Seems obsoloted by https://github.com/nextcloud/server/pull/36173

**Pull request closed without merging**